zookeeper-server/src/main/java/org/apache/zookeeper/ZooKeeper.java

```diff
@@ -2788,40 +2798,106 @@ public void setACL(final String path, List<ACL> acl, int version, StatCallback c
         ReplyHeader r = cnxn.submitRequest(h, request, response, wcb);
         if (r.getErr() != 0) {
             throw KeeperException.create(KeeperException.Code.get(r.getErr()),
-                    clientPath);
+                    path);
         }
+        if (outputStat != null) {
+            DataTree.copyStat(response.getStat(), outputStat);
+        }
+        updateNextPage(outputNextPage, response.getNextPageCzxid(), response.getNextPageCzxidOffset());
+
         return response.getChildren();
     }
 
+    /**
+     * Returns a list of all the children given the path.
+     * <p>
+     * The difference between this API and {@link #getChildren(String, boolean)} is:
+     * when there are lots of children and the network buffer exceeds {@code jute.maxbuffer},
+     * this API will fetch the children using pagination and be able to return all children;
+     * while {@link #getChildren(String, boolean)} will fail.
+     * <p>
+     * The final list of children returned is NOT strongly consistent with the server's data:
+     * the list might contain some deleted children if some children are deleted before the last page is fetched.
+     * <p>
+     * If the watch is true and the call is successful (no exception is thrown),
+     # a watch will be left on the node with the given path. The watch will be
+     * triggered by a successful operation that deletes the node of the given
+     * path or creates/deletes a child under the node.
+     *
+     * @param path the path of the parent node
+     * @param watch whether or not leave a watch on the given node
+     * @return a list of all children of the given path
+     * @throws KeeperException if the server signals an error with a non-zero error code.
+     * @throws InterruptedException if the server transaction is interrupted.
+     */
+    public List<String> getAllChildrenPaginated(String path, boolean watch)
+            throws KeeperException, InterruptedException {
+        PaginationNextPage nextPage = new PaginationNextPage();
+        Watcher watcher = watch ? watchManager.defaultWatcher : null;
+        List<String> childrenPaginated = getChildren(path, watcher, Integer.MAX_VALUE, 0L, 0, null, nextPage);
+        if (nextPage.getMinCzxid() == ZooDefs.GetChildrenPaginated.lastPageMinCzxid) {
+            return childrenPaginated;
+        }
+
+        Set<String> childrenSet = new HashSet<>(childrenPaginated);
+        do {
+            long minCzxid = nextPage.getMinCzxid();
+            // If a child with the same czxid is returned in the previous page, and then deleted
+            // on the server, the children not in the previous page but offset is less than czxidOffset
+            // would be missed in the next page. Use the last child in the previous page to determine whether or not
+            // the deletion case happens. If it happens, retry fetching the page starting from offset 0.
```

Comment:
I would suggest move this comment to the check at line 2872. Reader can understand why the check happens.

But for comment itself, I am a little bit confused. Is this doing consistent check? From what we agreed, we dont do consistent check, right?

Reply:
I tried to shorten the comments, but it seems it is not clear enough :)

Here is the case:
             // if a child with the same czxid is returned in the previous page, and then deleted
             // on the server, the starting offset for the next page should be shifted smaller accordingly.
             // If the next page still starts from czxidOffset, the children that not in the previous page
             // but their offset is less than czxidOffset, they would be missed.

For instance, there are 3 children with the same czxid 10: [0, 1, 2, 3]
In the 1st page, child [0, 1] is returned.
Then for the 2nd page, we'll start from offset 2 - suppose to get [2, 3]. But before that, the child 0 is deleted, only children [1, 2, 3] are left, then offsets of children 2 and 3 change from [2, 3] -> [1, 2]. When we continue with offset 2, we'll only get children [3]. Then child **2** is missed.

So the solution is, we start from the last fetched child 1. If deletion happens, we will not get the last fetched child 1 in the next page, then we know deletion happens, and retry from index 0.

Also appreciate @jiajunwang suggested the idea of using the last fetched child.

Reply:
So you mean adding will be appending to the end with larger zxid, it wont impact.

But still, as we discussed, even delete caused missing element, user should tolerate that, right? Otherwise, it is kind of doing consistent guarantee, but partial consistent guarantee. I think that is not a good idea.

Reply:
> So you mean adding will be appending to the end with larger zxid, it wont impact.

Right. Newly added child will have a larger zxid and so it will be at the end of the children list after sorting.

> But still, as we discussed, even delete caused missing element, user should tolerate that, right? Otherwise, it is kind of doing consistent guarantee, but partial consistent guarantee. I think that is not a good idea.

It's not doing consistency guarantee. We don't guarantee the deleted children on server would also be deleted in the final returned list - the deleted children could also be included in the final list.
But we have to make sure we don't miss any children, because the watcher is added in the final page call. If a child is missed to return, it won't be read until next watcher trigger. Then this is data miss, the client cannot tolerate this data miss.